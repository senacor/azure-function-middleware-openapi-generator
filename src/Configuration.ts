import {OpenAPIV3} from 'openapi-types';

export type Configuration = {
    apiDefinitions: ApiDefinitionConfiguration[]
}

export type ApiDefinitionConfiguration = {
    title: string;
    description?: string;
    servers?: OpenAPIV3.ServerObject[];
    outputFile: string;
    functionNameRegexToInclude?: string[];
    functionNameRegexToExclude?: string[]
}
