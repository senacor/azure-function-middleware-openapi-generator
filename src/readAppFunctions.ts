import * as azureFunctions from '@azure/functions';
import { HttpFunctionOptions } from '@azure/functions';
import * as azureFunctionMiddleware from '@senacor/azure-function-middleware';
import * as durableFunctions from 'durable-functions';
import { glob } from 'glob';
import { AnySchema } from 'joi';
import path from 'node:path';
import proxyquire from 'proxyquire';

export type AppFunctions = {
    httpFunctions: HttpFunction[];
};

export type HttpFunction = {
    name: string;
    options: HttpFunctionOptions;
    validations: HttpFunctionValidations;
};

export type HttpFunctionValidations = {
    hasJwtAuthorization: boolean;
    requestBodySchema?: AnySchema;
    requestQueryParamsSchema?: AnySchema;
    responseBodySchema?: Record<number, AnySchema>;
};

export function readAppFunctions(functionAppFilesPattern: string): AppFunctions {
    const httpFunctions: HttpFunction[] = [];

    const azureAppMock = {
        http: (name: string, options: HttpFunctionOptions) => {
            console.log(`Found http function ${name}`);
            httpFunctions.push({
                name,
                options,
                validations: extractDataFromMiddleware(options.handler as unknown as MiddlewareMock),
            });
        },
        activity: (name: string) => {
            console.log(`Found activity ${name}`);
        },
        orchestration: (name: string) => {
            console.log(`Found orchestration ${name}`);
        },
        timer: (name: string) => {
            console.log(`Found timer function ${name}`);
        },
        generic: (name: string) => {
            console.log(`Found generic function ${name}`);
        },
    };

    glob.sync(functionAppFilesPattern).forEach((file) => {
        proxyquire(path.resolve(file), {
            '@senacor/azure-function-middleware': {
                ...azureFunctionMiddleware,
                middleware: middlewareMock,
                jwtAuthorization: jwtAuthorizationMock,
                requestBodyValidation: requestBodyValidationMock,
                requestQueryParamsValidation: requestQueryParamsValidationMock,
                responseBodyValidation: responseBodyValidationMock,
            },
            '@azure/functions': {
                ...azureFunctions,
                app: azureAppMock,
            },
            'durable-functions': {
                ...durableFunctions,
                app: durableFunctionsAppMock(),
            },
        });
    });

    return {
        httpFunctions,
    };
}

type MiddlewareMock = {
    beforeExecution: ExecutionFunctionMock[];
    postExecution: ExecutionFunctionMock[];
};

type ExecutionFunctionMock = {
    type: 'jwtAuthorization' | 'requestBodyValidation' | 'requestQueryParamsValidation' | 'responseBodyValidation';
    schema?: AnySchema;
    schemaRecord?: Record<number, AnySchema>;
};

function middlewareMock(
    beforeExecution: ExecutionFunctionMock[],
    handler: unknown,
    postExecution: ExecutionFunctionMock[],
) {
    return {
        beforeExecution,
        postExecution,
    };
}

function jwtAuthorizationMock(): ExecutionFunctionMock {
    return {
        type: 'jwtAuthorization',
    };
}

function requestBodyValidationMock(schema: AnySchema): ExecutionFunctionMock {
    return {
        type: 'requestBodyValidation',
        schema,
    };
}

function requestQueryParamsValidationMock(schema: AnySchema): ExecutionFunctionMock {
    return {
        type: 'requestQueryParamsValidation',
        schema,
    };
}

function responseBodyValidationMock(schemaRecord: Record<number, AnySchema>): ExecutionFunctionMock {
    return {
        type: 'responseBodyValidation',
        schemaRecord,
    };
}

function extractDataFromMiddleware(handler: MiddlewareMock): HttpFunctionValidations {
    return {
        hasJwtAuthorization:
            handler.beforeExecution?.some((f: ExecutionFunctionMock) => f.type === 'jwtAuthorization') ?? false,
        requestBodySchema: handler.beforeExecution?.find(
            (f: ExecutionFunctionMock) => f.type === 'requestBodyValidation',
        )?.schema,
        requestQueryParamsSchema: handler.beforeExecution?.find(
            (f: ExecutionFunctionMock) => f.type === 'requestQueryParamsValidation',
        )?.schema,
        responseBodySchema: handler.postExecution?.find(
            (f: ExecutionFunctionMock) => f.type === 'responseBodyValidation',
        )?.schemaRecord,
    };
}

function durableFunctionsAppMock() {
    return {
        activity: (name: string) => {
            console.log(`Found durable activity ${name}`);
        },
        orchestration: (name: string) => {
            console.log(`Found durable orchestration ${name}`);
        },
    };
}
