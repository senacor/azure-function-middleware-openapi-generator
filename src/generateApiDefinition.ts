import {ApiDefinitionConfiguration, Configuration} from "./Configuration";
import {AppFunctions, HttpFunction} from "./readAppFunctions";
import fs from "node:fs";
import YAML from "yaml";
import {OpenAPIV3} from 'openapi-types';
import parse from "joi-to-json";
import {AnySchema} from "joi";

export function generateApiDefinitions(configuration: Configuration, appFunctions: AppFunctions, packageVersion: string) {
    configuration.apiDefinitions.forEach(config => {
        const apiDefinition = generateApiDefinition(config, appFunctions, packageVersion);
        writeApiDefinition(config, apiDefinition);
    });
}

export function generateApiDefinition(config: ApiDefinitionConfiguration, appFunctions: AppFunctions, packageVersion: string): OpenAPIV3.Document {
    const apiDefinition: OpenAPIV3.Document = {
        openapi: '3.0.2',
        info: {
            version: packageVersion,
            title: config.title,
            description: config.description,
        },
        servers: config.servers,
        paths: {},
        components: {
            securitySchemes: {
                FunctionKeyAuth: {
                    type: 'apiKey',
                    in: 'header',
                    name: 'x-functions-key'
                },
                BearerAuth: {
                    type: 'http',
                    scheme: 'bearer',
                    bearerFormat: 'JWT'
                },
            }
        }
    };

    appFunctions.httpFunctions
        .filter((httpFunction) => httpFunction.name && httpFunction.options.route)
        .filter((httpFunction) => shouldIncludeHttpFunctionInApiDefinition(config, httpFunction.name))
        .forEach((httpFunction) => {
            const fullRoute = `/${httpFunction.options.route}`;
            const operationObject: OpenAPIV3.OperationObject = {
                operationId: httpFunction.name,
                parameters: [...extractParametersFromRoute(fullRoute), ...generateQueryParametersDefinition(httpFunction)],
                requestBody: joiSchemaToOpenApi(httpFunction.validations.requestBodySchema),
                responses: generateResponses(httpFunction),
                security: generateSecurityForHttpFunction(httpFunction),
            };

            (httpFunction.options.methods ?? ['GET', 'POST'])
                .filter(method => method !== 'CONNECT')
                .map(method => OpenAPIV3.HttpMethods[method])
                .forEach((method) => {
                    if (!apiDefinition.paths[fullRoute]) {
                        apiDefinition.paths[fullRoute] = {};
                    }

                    apiDefinition.paths[fullRoute][method] = operationObject;
                });
        });

    return apiDefinition;
}

function shouldIncludeHttpFunctionInApiDefinition(config: ApiDefinitionConfiguration, name: string): boolean {
    if (!config.functionNameRegexToExclude && !config.functionNameRegexToInclude) {
        return true;
    }

    if (config.functionNameRegexToExclude?.some((regex) => name.match(regex)) === true) {
        return false;
    }

    return config.functionNameRegexToInclude?.some((regex) => name.match(regex)) === true;
}

function extractParametersFromRoute(route: string): OpenAPIV3.ParameterObject[] {
    return route
        .split('/')
        .filter((routePart) => routePart.startsWith('{') && routePart.endsWith('}'))
        .map((routePart) => {
            return {
                name: routePart.slice(1, routePart.length - 1),
                in: 'path',
                required: true,
                schema: {
                    type: 'string',
                },
            };
        });
}

function joiSchemaToOpenApi(schema?: AnySchema) {
    return schema ? parse(schema) : undefined;
}

function generateSecurityForHttpFunction(httpFunction: HttpFunction): OpenAPIV3.SecurityRequirementObject[] {
    const security = [];

    if (httpFunction.options.authLevel === 'function') {
        security.push({FunctionKeyAuth: []});
    }

    if (httpFunction.validations.hasJwtAuthorization) {
        security.push({BearerAuth: []})
    }

    return security;
}

function generateQueryParametersDefinition(httpFunction: HttpFunction): OpenAPIV3.ParameterObject[] {
    const queryOpenApiSchema = joiSchemaToOpenApi(httpFunction.validations.requestQueryParamsSchema);

    if (!queryOpenApiSchema) {
        return [];
    }

    if (queryOpenApiSchema.type !== 'object') {
        console.warn(`Request query parameter schema for ${httpFunction.name} is not valid.`);
        return [];
    }

    const queryParams: OpenAPIV3.ParameterObject[] = [];

    Object.entries(queryOpenApiSchema.properties).forEach(([key, value]) => {
        const {description, ...remainingSchema} = value as OpenAPIV3.NonArraySchemaObject;

        queryParams.push({
            name: key,
            in: 'query',
            required: queryOpenApiSchema.required?.includes(key) ?? false,
            description: description,
            schema: remainingSchema,
        })
    })

    return queryParams;
}

function generateResponses(httpFunction: HttpFunction): OpenAPIV3.ResponsesObject {
    if (!httpFunction.validations.responseBodySchema) {
        return {
            default: {
                description: 'ok',
            },
        }
    }

    const responses: OpenAPIV3.ResponsesObject = {};

    if (httpFunction.validations.responseBodySchema) {
        Object.entries(httpFunction.validations.responseBodySchema).forEach(([status, schema]) => {
            responses[status] = {
                description: `${status} response`,
                content: {
                    'application/json': {
                        schema: joiSchemaToOpenApi(schema)
                    }
                }
            }
        })
    }

    return responses;
}

function writeApiDefinition(config: ApiDefinitionConfiguration, apiDefinition: OpenAPIV3.Document) {
    fs.writeFileSync(config.outputFile, config.outputFile.endsWith('.json') ? JSON.stringify(apiDefinition, null, 2) : YAML.stringify(apiDefinition), {
        encoding: 'utf-8',
    });
}