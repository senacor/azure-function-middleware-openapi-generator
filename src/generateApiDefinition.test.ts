import {generateApiDefinition} from "./generateApiDefinition";
import {ApiDefinitionConfiguration} from "./Configuration";
import {AppFunctions} from "./readAppFunctions";
import Joi from "joi";

describe('generateApiDefinition should', () => {
    const config: ApiDefinitionConfiguration = {
        title: 'Test API',
        outputFile: 'test.yaml'
    }

    test('generate api definition for http endpoint without validations', () => {
        const appFunctions: AppFunctions = {
            httpFunctions: [
                {
                    name: 'http-get-test',
                    options: {
                        authLevel: 'function',
                        methods: ['GET'],
                        route: 'test/{id}',
                        handler: async () => ({status: 200}),
                    },
                    validations: {
                        hasJwtAuthorization: false,
                    }
                }
            ]
        }

        const result = generateApiDefinition(config, appFunctions, '1.0.0');

        expect(result).toMatchSnapshot();
    });

    test('generate api definition for http endpoint with url and query params', () => {
        const appFunctions: AppFunctions = {
            httpFunctions: [
                {
                    name: 'http-get-test',
                    options: {
                        authLevel: 'function',
                        methods: ['GET'],
                        route: 'person/{personId}/order/{orderId}/item',
                        handler: async () => ({status: 200}),
                    },
                    validations: {
                        hasJwtAuthorization: false,
                        requestQueryParamsSchema: Joi.object({
                            status: Joi.string().valid('active', 'expired').required(),
                        })
                    }
                }
            ]
        }

        const result = generateApiDefinition(config, appFunctions, '1.0.0');

        expect(result).toMatchSnapshot();
    });

    test('generate api definition for http endpoint with request and response body validation', () => {
        const appFunctions: AppFunctions = {
            httpFunctions: [
                {
                    name: 'http-post-person',
                    options: {
                        authLevel: 'anonymous',
                        methods: ['POST'],
                        route: 'person',
                        handler: async (req) => {
                            const requestBody = await req.clone().json() as { name: string };
                            return {status: 200, jsonBody: {name: requestBody.name}};
                        },
                    },
                    validations: {
                        hasJwtAuthorization: true,
                        requestBodySchema: Joi.object({
                            name: Joi.string().required(),
                        }),
                        responseBodySchema: {
                            200: Joi.object({
                                name: Joi.string().required(),
                            }),
                        }
                    }
                }
            ]
        }

        const result = generateApiDefinition(config, appFunctions, '1.0.0');

        expect(result).toMatchSnapshot();
    })
})