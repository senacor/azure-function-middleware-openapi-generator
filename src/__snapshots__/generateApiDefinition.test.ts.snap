// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateApiDefinition should generate api definition for http endpoint with request and response body validation 1`] = `
{
  "components": {
    "securitySchemes": {
      "BearerAuth": {
        "bearerFormat": "JWT",
        "scheme": "bearer",
        "type": "http",
      },
      "FunctionKeyAuth": {
        "in": "header",
        "name": "x-functions-key",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "description": undefined,
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.2",
  "paths": {
    "/person": {
      "post": {
        "operationId": "http-post-person",
        "parameters": [],
        "requestBody": {
          "additionalProperties": false,
          "properties": {
            "name": {
              "type": "string",
            },
          },
          "required": [
            "name",
          ],
          "type": "object",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "name",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "200 response",
          },
        },
        "security": [
          {
            "BearerAuth": [],
          },
        ],
      },
    },
  },
  "servers": undefined,
}
`;

exports[`generateApiDefinition should generate api definition for http endpoint with url and query params 1`] = `
{
  "components": {
    "securitySchemes": {
      "BearerAuth": {
        "bearerFormat": "JWT",
        "scheme": "bearer",
        "type": "http",
      },
      "FunctionKeyAuth": {
        "in": "header",
        "name": "x-functions-key",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "description": undefined,
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.2",
  "paths": {
    "/person/{personId}/order/{orderId}/item": {
      "get": {
        "operationId": "http-get-test",
        "parameters": [
          {
            "in": "path",
            "name": "personId",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "orderId",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "description": undefined,
            "in": "query",
            "name": "status",
            "required": true,
            "schema": {
              "enum": [
                "active",
                "expired",
              ],
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "default": {
            "description": "ok",
          },
        },
        "security": [
          {
            "FunctionKeyAuth": [],
          },
        ],
      },
    },
  },
  "servers": undefined,
}
`;

exports[`generateApiDefinition should generate api definition for http endpoint without validations 1`] = `
{
  "components": {
    "securitySchemes": {
      "BearerAuth": {
        "bearerFormat": "JWT",
        "scheme": "bearer",
        "type": "http",
      },
      "FunctionKeyAuth": {
        "in": "header",
        "name": "x-functions-key",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "description": undefined,
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.2",
  "paths": {
    "/test/{id}": {
      "get": {
        "operationId": "http-get-test",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "default": {
            "description": "ok",
          },
        },
        "security": [
          {
            "FunctionKeyAuth": [],
          },
        ],
      },
    },
  },
  "servers": undefined,
}
`;
